### PR TITLE
refactor: Consume from the eclipse junit runner

### PR DESCRIPTION
**Pull request by jdneo**

### CODE REVIEW

File: src/runners/baseRunner/BaseRunner.ts [hunk 1]
```diff
@@ -180,6 +182,25 @@ export abstract class BaseRunner implements ITestRunner {
         return commandParams;
     }
 
+    protected getJavaEncoding(): string {
```
Comment (jdneo) [inline src/runners/baseRunner/BaseRunner.ts]: These two methods are just changed modifier from `private` to `protected` and changed the location up

File: src/runners/junit4Runner/Junit4Runner.ts [hunk 2]
```diff
@@ -1,14 +1,129 @@
...
+                mainClass: this.arguments.mainClass,
+                projectName: this.arguments.projectName,
+                cwd: this.config ? this.config.workingDirectory : undefined,
+                classPaths: this.arguments.classpath,
+                modulePaths: this.arguments.modulepath,
+                args: this.arguments.programArguments,
+                vmArgs: this.arguments.vmArguments.join(' '),
+                encoding: this.encoding,
+                env,
+                console: 'internalConsole',
```
Comment (Eskibear) [inline src/runners/junit4Runner/Junit4Runner.ts]: Cool

File: src/runners/junit4Runner/Junit4Runner.ts [hunk 3]
```diff
@@ -1,14 +1,129 @@
...
+                noDebug: this.isDebug ? false : true,
```
Comment (testforstephen) [inline src/runners/junit4Runner/Junit4Runner.ts]: !this.isDebug

File: src/runners/junit4Runner/JUnit4RunnerResultAnalyzer.ts [hunk 4]
```diff
@@ -1,55 +1,89 @@
...
+import { ITestResultDetails, TestStatus } from '../models';
 
 export class JUnit4RunnerResultAnalyzer extends BaseRunnerResultAnalyzer {
 
+    private currentTestItem: string;
+    private traces: string;
+    private isRecordingTraces: boolean;
+
+    public analyzeData(data: string): void {
+        const lines: string[] = data.split(/\r?\n/);
```
Comment (testforstephen) [inline src/runners/junit4Runner/JUnit4RunnerResultAnalyzer.ts]: how did you guarantee that you will receive an entire line of data at a time？

File: src/runners/junit4Runner/Junit4Runner.ts [hunk 5]
```diff
@@ -1,14 +1,129 @@
...
+            if (this.config && this.config.vmargs) {
+                this.arguments.vmArguments.push(...this.config.vmargs.filter(Boolean));
+            }
+            let env: {} = process.env;
+            if (this.config && this.config.env) {
+                env = { ...process.env, ...this.config.env };
+            }
+
+            let data: string = '';
+            this.server.on('connection', (socket: Socket) => {
```
Comment (testforstephen) [inline src/runners/junit4Runner/Junit4Runner.ts]: When to close the server?

File: src/runners/junit4Runner/Junit4Runner.ts [hunk 6]
```diff
@@ -1,14 +1,129 @@
...
+                socket.on('data', (buffer: Buffer) => {
+                    data = data.concat(iconv.decode(buffer, this.encoding));
+                    const index: number = data.lastIndexOf(os.EOL);
+                    if (index >= 0) {
+                        testResultAnalyzer.analyzeData(data.substring(0, index + os.EOL.length));
+                        data = data.substring(index + os.EOL.length);
+                    }
+                });
+
+                socket.on('close', (had_error: boolean) => {
```
Comment (testforstephen) [inline src/runners/junit4Runner/Junit4Runner.ts]: The logic duplicated with BaseRunner.

File: src/runners/ITestRunner.ts [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+import { ISearchTestItemParams, ITestItem } from '../protocols';
 import { IExecutionConfig } from '../runConfigs';
 import { ITestResult } from './models';
 
 export interface ITestRunner {
-    setup(tests: ITestItem[], isDebug: boolean, server: Server, config?: IExecutionConfig): Promise<void>;
-    run(): Promise<ITestResult[]>;
+    setup(tests: ITestItem[], isDebug: boolean, config?: IExecutionConfig, searchParam?: ISearchTestItemParams): Promise<DebugConfiguration>;
+    run(launchConfiguration: DebugConfiguration): Promise<ITestResult[]>;
     cleanUp(isCancel: boolean): Promise<void>;
```
Comment (testforstephen) [inline src/runners/ITestRunner.ts]: => tearDown

File: src/runners/baseRunner/BaseRunner.ts [hunk 8]
```diff
@@ -1,186 +1,182 @@
...
+            request: 'launch',
+            mainClass: this.runnerMainClassName,
+            projectName: tests[0].project,
+            cwd: config ? config.workingDirectory : undefined,
+            classPaths,
+            args: this.getApplicationArgs(config),
+            vmArgs: this.getVmArgs(config),
+            encoding: getJavaEncoding(Uri.parse(tests[0].location.uri), config),
+            env,
+            noDebug: !isDebug ? true : false,
```
Comment (testforstephen) [inline src/runners/baseRunner/BaseRunner.ts]: no need of ternary expression. Use `!isDebug` directly.

**Pull request merged**